python/dgl/graph.py

```diff
@@ -988,6 +1007,9 @@ def add_nodes(self, num, data=None):
                 [1., 1., 1., 1.],
                 [1., 1., 1., 1.]])
         """
+        if self.is_readonly:
+            raise DGLError("Readonly graph. Mutation is not allowed.")
+
```

Comment:
Remember to modify `batch_num_nodes` and `batch_num_edges`. The simplest way is to flatten the graph.

Reply:
So I suggest to let `batch_num_nodes` and `batch_num_edges` to depend on `number_of_nodes` and `number_of_edges` for single graphs.